src/core/processing/queue.ts

```diff
@@ -0,0 +1,265 @@
+ * drifts are then the actual elements in the queue.
+ *
+ * In addition, the decaying deque runs a timer that purges old elements from
+ * the queue. This period of time is determined by the `handlerTimeout`.
+ *
+ * When a task completes or exceeds its timeout, the corresponding drift is
+ * removed from the queue. As a result, only drifts with pending tasks are
+ * contained in the queue at all times.
+ *
+ * When a tasks completes with failure (`reject`s or exceeds the timeout), the
+ * respective handler (`catchError` or `catchTimeout`) is called.
+ *
+ * The decaying deque has its name from the observation that new elements are
+ * appended to the tail, and the old elements are removed at arbitrary positions
+ * in the queue whenever a task completes, hence, the queue seems to _decay_.
+ */
+export class DecayingDeque<E> {
+  /**
+   * Number of drifts in the queue. Equivalent to the number of currently
+   * pending tasks.
+   */
+  private length: number = 0
+  /** Head element (oldest), `null` iff the queue is empty */
+  private head: Drift<E> | null = null
+  /** Tail element (newest), `null` iff the queue is empty */
+  private tail: Drift<E> | null = null
+
+  /**
+   * Number of currently pending tasks that we strive for (`add` calls will
+   * resolve only after the number of pending tasks falls below this value.
+   *
+   * In the context of `telegraf`, it is possible to `await` calls to `add` to
+   * determine when to fetch more updates.
+   */
+  private readonly concurrency: number
+  /**
+   * Timer that waits for the head element to time out, will be rescheduled
+   * whenever the head element changes. It is `undefined` iff queue is empty.
+   */
+  private timer: NodeJS.Timeout | undefined
+  /**
+   * List of subscribers that wait for the queue to be empty. All functions in
+   * this array will be called as soon as new capacity is available, i.e. the
+   * number of pending tasks falls below `concurrency`.
+   */
+  private subscribers: Array<(capacity: number) => void> = []
+
+  /**
+   * Creates a new decaying queue with the given parameters.
+   *
+   * @param handlerTimeout max period of time for a task
+   * @param worker task generator
+   * @param concurrency `add` will return only after the number of pending tasks fell below `concurrency`. `false` means `1`, `true` means `Infinity`, numbers below `1` mean `1`
+   * @param catchError error handler, receives the error and the source element
+   * @param catchTimeout timeout handler, receives the source element and the promise of the task
+   */
+  constructor(
+    private readonly handlerTimeout: number,
+    private readonly worker: (t: E) => Promise<void>,
+    concurrency: boolean | number,
+    private readonly catchError: (err: unknown, elem: E) => Promise<void>,
+    private readonly catchTimeout: (t: E, task: Promise<void>) => void
+  ) {
+    if (concurrency === false) this.concurrency = 1
+    else if (concurrency === true) this.concurrency = Infinity
```

Comment:
Very unsafe, will confirm updates prematurely and fetch more even when bot is overloaded.

```suggestion
    else if (concurrency === true) this.concurrency = 100
```

Reply:
My initial iteration had 100 as default, but this is a problem when the load goes beyond that value for webhooks. Then suddenly with too much load the `add` calls would be slowed down until Telegram re-sends updates, thus making deduplication necessary again. `Infinity` is the only choice where this does not happen, even though it looks odd.

Reply:
So, `100` is incorrect for webhooks, while `Infinity` is incorrect for polling?

Reply:
`1` is correct for polling, larger values may lead to calls to `getUpdates` before all middleware has completed. (One may choose to increase the value and decrease reliability to improve performance.)

`Infinity` is correct for webhooks, lower values may lead to duplicate updates being received. (I don't think it makes much sense to decrease the value, but perhaps someone else knows a use-case.)